ehcache-impl/src/main/java/org/ehcache/impl/internal/store/tiering/TieredStore.java

```diff
@@ -356,13 +357,19 @@ public Map<K, ValueHolder<V>> bulkCompute(Set<? extends K> keys, Function<Iterab
   }
 
   @Override
-  public Map<K, ValueHolder<V>> bulkComputeIfAbsent(Set<? extends K> keys, Function<Iterable<? extends K>, Iterable<? extends Map.Entry<? extends K, ? extends V>>> mappingFunction) throws StoreAccessException {
+  public Map<K, Store.ValueHolder<V>> bulkComputeIfAbsent(Set<? extends K> keys, Function<Iterable<? extends K>, Iterable<? extends Map.Entry<? extends K, ? extends V>>> mappingFunction) throws StoreAccessException {
+
     try {
-      return authoritativeTier.bulkComputeIfAbsent(keys, mappingFunction);
-    } finally {
-      for (K key : keys) {
-        cachingTier().invalidate(key);
-      }
+      return cachingTier().bulkGetOrComputeIfAbsent(keys, missingKeys -> {
+        try {
+          return authoritativeTier.bulkComputeIfAbsentAndFault(missingKeys, mappingFunction);
+        } catch (StoreAccessException cae) {
+          throw new StorePassThroughException(cae);
+        }
+      });
+    } catch (StoreAccessException ce) {
+      handleStoreAccessException(ce);
+      return Collections.emptyMap();
```

Comment:
This return isn't right per the contract but will never happen because the method call above always throws. We should instead modify the method so that we can use it in the same manner in all places. I'm thinking either introduce a generic for the return type so that it can be inferred, or convert it to a `throw ...` and return an exception from the method.

Reply:
Can you elaborate this more?

Reply:
Returning `Collections.emptyMap()` here is wrong... but the line is unreachable because `handleStoreAccessException(...)` always throws. The other uses of `handleStoreAccessException(...)` get around this problem by having:
```
return handleStoreAccessException(e)
```
because its declared to return the right type (even though it doesn't return anything). You should be able to adapt the method to allow you to use it in that way here too.

Reply:
modified the method signature to generic return type to accomodate both types of return type - ValueHolder<V> & Map<K, Store.ValueHolder<V>>

```
-  private ValueHolder<V> handleStoreAccessException(StoreAccessException ce) throws StoreAccessException
+  private <R> R handleStoreAccessException(StoreAccessException ce) throws StoreAccessException
```